synphot/tests/test_observation.py

```
@@ -439,3 +440,136 @@ def test_countrate_neg_leak():
+        Cambridge University Press)
+    """
+    t = 300 * u.s
+    snr = 342
+    gain = 5 * (u.ct / u.adu)
+    countrate = 24013 * u.adu * gain / t
+    npix = 1 * u.pixel
+    n_background = 200 * u.pixel
+    background_rate = 620 * (u.adu / u.pixel) * gain / t
+    darkcurrent_rate = 22 * (u.ct / u.pixel / u.hr)
+    readnoise = 5 * (u.ct / u.pixel)
+
+    result = exptime_from_ccd_snr(snr, countrate, npix=npix,
+                                  n_background=n_background,
+                                  background_rate=background_rate,
+                                  darkcurrent_rate=darkcurrent_rate,
+                                  readnoise=readnoise, gain=gain)
+
+    assert_quantity_allclose(result, t, atol=1 * u.s)
+
+
+def test_t_exp_analytic():
+    """
+    A test to check that the analytic method in
+    observation.exptime_from_ccd_snr is done correctly.
+    """
+    snr_set = 50
+    countrate = 1000 * (u.ct / u.s)
+    npix = 1 * u.pixel
+    background_rate = 100 * (u.ct / u.pixel / u.s)
+    darkcurrent_rate = 5 * (u.ct / u.pixel / u.s)
+    readnoise = 1 * (u.ct / u.pixel)
+
+    t = exptime_from_ccd_snr(snr_set, countrate, npix=npix,
+                             background_rate=background_rate,
+                             darkcurrent_rate=darkcurrent_rate,
+                             readnoise=readnoise)
+
+    # if t is correct, ccd_snr() should return snr_set:
+    snr_calc = ccd_snr(countrate * t,
+                       npix=npix,
+                       background=background_rate * t,
+                       darkcurrent=darkcurrent_rate * t,
+                       readnoise=readnoise)
+
+    assert_quantity_allclose(snr_calc, snr_set,
+                             atol=0.5 * u.Unit(""))
+
+
+def test_snr_with_countrate():
+    """
+    A test to make sure `ccd_snr` works with the units that
+    `countrate` returns.
+    """
+    source_spec = SourceSpectrum(BlackBodyNorm1D)
```

For sanity check, we can set this up so that the observation would produce appox. 1 counts/s.
```suggestion
    area = 1 * u.m ** 2
    bp = SpectralElement(Const1D)
    source_spec = SourceSpectrum(BlackBodyNorm1D)
    source_spec = source_spec.normalize(
        1 * u.ct, bp, force='extrap', wavelengths=source_spec.waveset, area=area)
```

---

Ah I'm glad you suggested this because it didn't pass! And it's because I set the default gain to 1 ct/adu and the default ad_err to some other number, so for very small counts (like 1 in this test) the gain error significantly changes the snr, even if it isn't specified by the user. So to keep up with the "idealized" snr that we represent by default, I changed the default gain to 0 such that its error is ignored unless specified.

---

Well, if 1 count is an edge case that requires gain of zero, you could also consider normalizing to higher count values. What do you think?

---

Nah, I think it's good to test even these edge cases. Including the gain error by default doesn't make too much sense anyway since in most cases it will probably be ignored.